### ISSUE TITLE
Support base58 encoding private keys (WIF)

### ISSUE THREAD

**Issue by stevenroose** [rust-bitcoin/rust-bitcoin#164]
There is an official [base58 encoding for private keys](https://en.bitcoin.it/wiki/Base58Check_encoding#Encoding_a_private_key) (aka wallet import format). Currently neither base58 can be accessed publicly neither does the `Privkey` type support serializing to WIF/base58.

**Comment by bodymindarts** [comment]
Just looking through the code and want to clarify.
Afaict the `Display` and `FromStr` traits that `PrivKey` [implement](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/util/privkey.rs#L97-L143) do serialise to/from WIF format.
What is missing here?

**Comment by stevenroose** [comment]
@bodymindarts Was not aware of that. It's not trivial though that the Display impl for PrivKey is WIF and one should also not rely on that to be true forever. I would suggest having `to_wif` and `from_wif` methods and using those in `Display` and `FromStr` instead. 

But I think @dongcarl might be doing something related to this.

**Comment by dongcarl** [comment]
@bodymindarts you are absolutely right.

I'm working on refactoring the `key` module, so we'll make sure to address this. `to_wif` and `from_wif` should be implemented. And we'll make sure to mention this in the documentation.

**Comment by apoelstra** [comment]
@stevenroose the Display format will always be WIF because that's what is used by `to_string()`. I don't understand why we need new methods to convert private keys to and from strings.

The same behaviour is true for addresses etc.

**Comment by stevenroose** [comment]
I think addresses are different. "Address" inherently means "a text representation of something" (i.e. pubkey, script, ...) so since this is inherently a text representation, it's obvious that the `Display` trait gives you that text representation. 

The `PrivateKey` type on the otherhand is *not* "a text representation of a private key". Instead, it's just a private key. This means that it's not obvious that the `Display` trait will give you WIF. Since WIF is just *a possible text representation of a private key*. I think the most common representation of private keys in general, even in Bitcoin applications, is plain 32-byte hex.

Basically what I'm saying is that the contract of the `Display` trait doesn't assure me I will get WIF instead of hex. And thus that's subject to change in the future.

**Comment by dpc** [comment]
I agree with @stevenroose  that explicit `to_wif` and `from_wif` are better: more explicit, self-documenting, easy to search/grep for.

Some Rust crates dealing with secret keys and other sensitive material do not implement `Display` and `Debug` on them at all for a simple reason - to make it harder to make mistakes of accidentally leaking private keys.

Eg. key might be embedded into a bigger struct, with `#[derive(Display)]`, and that struct accidentally added to logging by an unaware developer. The Rust-way is designing APIs that are hard to misuse.

**Comment by apoelstra** [comment]
Also, getting rid of `Display` and `FromStr` would make interop with script descriptors much more difficult because suddenly we can't be generic over stringified keys.

**Comment by dpc** [comment]
> AFAIK there is no such thing as `#[derive(Display)]`.

Oh. Right. But `Debug` is. 

> Without a Display impl it is impossible to encode a private key without allocating a new string or introducing a serde dependency.

Isn't it possible to just expose the same method manually?

```
impl PrivateKey {
   fn fmt_as_wif(&self, f: &mut fmt::Formatter) -> fmt::Result {
      /// ...
  }
}
```

> Also, getting rid of Display and FromStr would make interop with script descriptors much more difficult because suddenly we can't be generic over stringified keys.

I'm not entierely sure about this, but it seems to me you can always have an internal trait, just like `Display` that you `impl<T: Display> JustLikeDisplay for T` and `impl JustLikeDisplay PrivateKey`, and things should work roughtly as they used to.

I might be missing something here though.

**Comment by apoelstra** [comment]
We can certainly make the `Debug` impl output `<private key data>` or something like that because there's a good chance it will end up being used by mistake.

`Display` is a standard way of stringifying objects which is needed in generic programming contexts and which is never used implicitly. People are not going to use some rust-bitcoin-specific `JustLikeDisplay` trait that has exactly the same properties except that it's nonstandard and nondiscoverable.

**Comment by apoelstra** [comment]
If it could be made ergonomic (e.g. how would `FromStr` be made to work?) I'd be ok with `PrivKeyWifSerialization`. But it's not obvious to me how to do that.

**Comment by dpc** [comment]
```
// deserialization
let privkey : PrivKey = PrivKeyWifSerialization::from_str(s).into();
let privkey = PrivKeyWifSerialization::from_str(s).into_privkey();

// serialization
println!("{}", privkey.to_wif());
``` 

To have both non-cloning serialization, and non-cloning (move) on deserialization, I guess `PrivKeyWifSerialization` would have to contain `Cow<'a,  PrivKey>`, but seems doable.

**Comment by apoelstra** [comment]
FWIW, cloning a privkey is super cheap, it's only 36 bytes.

**Comment by stevenroose** [comment]
Let's not go even deeper down the rabbit hole..

What's wrong with just having `to_wif` and `from_wif` explicitly and reuse them in `Display` and `FromStr` as long as WIF is the default string serialization method for private keys. That way a user that just wants to print private keys no matter what can do it, and a user that really needs to make sure he gets WIF even after future rust-bitcoin updates can use the explicit methods.

And if we're really concerned about the performance of allocating a string in `Display`, we can have `fmt_wif`. :)

**Comment by stevenroose** [comment]
I'd be happy to make a PR right now :)

**Comment by sgeisler** [comment]
Can this issue be closed since #178 was merged?

**Comment by apoelstra** [comment]
Yes.
